### PR TITLE
Add check for existing crossrefs when importing entries into a library 

### PR DESCRIPTION
**Pull request by tom-pjobrien**
- Added a check to ensure citation keys generate correctly when importing an already-crossreferenced entry into a library

Before:
image

After:
image

This PR fixes https://github.com/JabRef/jabref/issues/11136

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java
Comment (InAnYan) [inline src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:307]: As an exercise in Java programming you can eliminate the outer loop using Java Streams API. Like this: `entries.stream()...<logic to filter a parent entry>...findOne()`, and you will have an `Optional` as the result.

But, of course, eliminating one loop doesn't mean you reduced the complexity of the function, it's still $O(n^2)$ with streams

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java
Comment (InAnYan) [inline src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:312]: This logic with years is wrong.

You see, there are citation key patterns in JabRef, and by default it's something like `[author][year]`, so every generated citation key will consists of the last name of the first author and a year of paper. It's the default, and that is why you added year logic there. But, for example, I may decide not to include a year, maybe I decided to make citation pattern to look like a title of the paper.

**Review by InAnYan — CHANGES_REQUESTED** [review]
I've left some small comments on your PR, but probably you also need a review from main maintainers, as I haven't fully understood the original issue

**Comment by tom-pjobrien** [comment]
Thanks for the comments @InAnYan, I'll take a look through and update the key generation logic.

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java
Comment (ThiloteE) [reply src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:312]: It is true. In JabRef one can customize the citationkey patterns extensively. See https://docs.jabref.org/setup/citationkeypatterns

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:312]: We have the class CitationKeyGenerator who does this all under the hood e..g generateKey method

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java
Comment (calixtus) [reply src/main/java/org/jabref/gui/externalfiles/ImportHandler.java]: For the mental closure: its obsolete now with the call to the citatationkeygenerator, but maybe parallelstreams would have helped with execution speed here.

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java
Comment (koppor) [inline src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:204]: Why the list? What happens on a second import of other entries? Will the citation keys of the previous entries be replaced?

**Review by koppor — CHANGES_REQUESTED** [review]
Strange change...

Did you test it?

**Comment by tom-pjobrien** [comment]
> Strange change...
> 
> Did you test it?

I tested this change manually with a few different inputs and variations (e.g importing crossreferenced entries at the same time in different orders, importing them in different imports etc.). It worked with all key patterns I tried as well as for general use. 

I used a list since in practice entries are imported and have keys generated one at a time, so adding them to a list as they are imported and then generating keys on all of them at the end is the best way of ensuring crossref relationships are respected in terms of citation keys. Currently if a second import occurs the list will be maintained, but if old keys should not be regenerated that's an easy fix as well.

**Review by koppor — CHANGES_REQUESTED** [review]
Please add two tests to org.jabref.gui.externalfiles.ImportHandlerTest

One where the `List` of entries has the cross-refernced entry first, the other where it comes second.

Ensure that the citationkeygenerator is active. (This is a harder one, because the test currently does not have the citation key generator enabled).

This is crucial to automatically check that the poster's issue is resolved:

image

---

I would suggest to remove the variable `finalizedEntries` and directly pass `entries` to `generateKeys` as I don't see any additional value of this variable.

**Comment by tom-pjobrien** [comment]
I can absolutely work on writing some tests. 

> I would suggest to remove the variable `finalizedEntries` and directly pass `entries` to `generateKeys` as I don't see any additional value of this variable.

I believe that passing `entries` to `generateKeys` is what was causing the original issue. `importCleanedEntries` imports entries one at a time. For example if I used your above reproducer, the first time `importCleanedEntries` is called it will only generate a key for the first entry, and then the second entry the next time it is called etc.:
image
Essentially `entries` will always be a size-1 list of a single entry. If a child is imported before a parent this will then lead to its citekey being incorrectly generated. Using a list which collects all entries will 're-generate' the child entry's key with respect to the parent entry. After an import is completed the list is cleared so it does not effect future imports. If there is a better way to implement this please to let me know.

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java
Comment (koppor) [inline src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:80]: rename list to `addedEntries`. `finalized` is a strange word.

Also remove `final` keyword, because the list itself is modified; thus `final` is misleading.

**Comment by koppor** [comment]
I think, this change is required because of our update at #10741 to org.jabref.gui.externalfiles.ImportHandler#importEntriesWithDuplicateCheck -- now, entries get added one by one and not as complete list. - This PR reassembles the complete list. @Siedlerchr I don't see a better solution for now. Do you?

**Comment by tom-pjobrien** [comment]
I have made the requested changes and added tests, so everything is now working correctly. However, I have noticed something unusual. The previous case which worked correctly before my changes (parent first, then child) no longer works correctly with a single call of generateKeys, as it seems like some other commit has caused issues with crossref inheritance. Calling generateKeys a first time across all entries before they are added to the database (in importEntriesWithDuplicateCheck) resolves this, so it's a small issue, but I just wanted to point it out.

File: src/main/java/org/jabref/gui/externalfiles/ImportHandler.java
Comment (koppor) [inline src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:426]: Wrong indent.

**Comment by koppor** [comment]
> The previous case which worked correctly before my changes (parent first, then child) no longer works correctly with a single call of generateKeys, as it seems like some other commit has caused issues with crossref inheritance. Calling generateKeys a first time across all entries before they are added to the database (in importEntriesWithDuplicateCheck) resolves this, so it's a small issue, but I just wanted to point it out.

I do not understand. Is there a test case for it? - You moved `generateKeys(addedEntries);` down - I haven't checked that method if it also updates the crossrefs.

**Comment by tom-pjobrien** [comment]
There is a test case for it and it passes. I had to make some small adjustments to my implementation, but my tests pass and everything is working as expected.

**Comment by koppor** [comment]
> There is a test case for it and it passes. I had to make some small adjustments to my implementation, but my tests pass and everything is working as expected.

Any hint which concrete test case it is?

I need more time to understand the "small issue" and what needs to be done.

---

I understand: With your PR: Something is broken which is not broken before.

**Comment by koppor** [comment]
To  review:

1. Try if https://github.com/JabRef/jabref/issues/11136#issuecomment-2057752405 works now
2. Swap the entries of https://github.com/JabRef/jabref/issues/11136#issuecomment-2057752405 and try again (in a new library)
3. Switch to another library try pasting other entries - does that still work?


Maybe adjust the test text of https://github.com/JabRef/jabref/issues/11136#issuecomment-2057752405 - it seems to be incomplete.

- Starting with an empty library
- Configuration of citation key generator

**Comment by tom-pjobrien** [comment]
> Any hint which concrete test case it is?

The test case is `importCrossRefParentFirstTest`.

> But the "smal lissue" described at https://github.com/JabRef/jabref/pull/11922#issuecomment-2423652274 is unclear to me.

It was originally the case that when a parent was imported before a child it used to work before I made any changes, but this stopped working. I made an adjustment to my code so it works again, by initially calling `generateKeys` across all entries in `importEntriesWithDuplicateCheck` to establish crossref keys. Each entry is then inserted one at a time in `importCleanedEntries` where its key is re-generated with duplicate checks + automatic fields in mind. So there is no issue at all anymore, I just wanted to explain my changes. Apologies for the confusion.

> Try if https://github.com/JabRef/jabref/issues/11136#issuecomment-2057752405 works now
> Swap the entries of https://github.com/JabRef/jabref/issues/11136#issuecomment-2057752405 and try again (in a new library)
> Switch to another library try pasting other entries - does that still work?

I have done all of the above and everything is working as expected. I also tried a number of key generation patterns ([year][auth], [title][year] etc.) as well as using my own test data. The parent's year is used correctly for the child's key no matter the order of import/state of the libraries.

**Comment by koppor** [comment]
> I made an adjustment to my code so it works again, by initially calling `generateKeys` across all entries in `importEntriesWithDuplicateCheck` to establish crossref keys.

Strange change. The entries pasted should have working referencing keys. If it only works after an update, something is wrong probably somewhere else.

**Comment by Siedlerchr** [comment]
Please be aware that you should only generate keys if the preferences are set but I think that generateCitaitonKeys already has a check

**Comment by koppor** [comment]
> Please be aware that you should only generate keys if the preferences are set but I think that generateCitaitonKeys already has a check

Already handled:

```java
    private void generateKeys(List<BibEntry> entries) {
        if (!preferences.getImporterPreferences().isGenerateNewKeyOnImport()) {
            return;
        }

```

Our method name is not that good.

**Review by koppor — CHANGES_REQUESTED** [review]
Does not work at all.

image

The `Crossref` value is NOT updated on key generation. This needs to be fixed. 

`org.jabref.logic.citationkeypattern.CitationKeyGenerator#generateAndSetKey` needs to be modifiered to return a `List<FieldChange>`. (List is empty on no change)

org.jabref.logic.citationkeypattern.CitationKeyGenerator#generateKey needs to be adpated to go through `this.database.getEntries()` and check `crossref` of each entry. If it is the old key: replace with new one and break.

You can try out with following entries - there is no `year` used for citaton key generation on paste for the second article.

```bibtex
@Article{parent,
  author   = {first-author},
  year     = {2024},
}

@Article{,
  author   = {third-best},
  title    = {title},
  crossref = {parent},
}
```
Thus,

**Comment by koppor** [comment]
I think, citaton keys are properly updated using the `org.jabref.model.database.KeyChangeListener`.

**Comment by koppor** [comment]
Closing this issue due to inactivity :zzz: 

Please ping us if you intend to resume work on this one.

**Pull request closed without merging**
